### DISCUSSION TITLE
How do we compile the code into a standalone binary?

### DISCUSSION THREAD

**Discussion by geocine** [NicoNex/tau#35] [General]
I am currently in the process of learning compilers and interpreters and been following Thorsten Ball's books. I saw this repository and @prologic repository , which  are both great complementary learning resources. 

I have been reading about what LLVM is, but currently its beyond me. I am definitely willing to shell out money to learn the end to end process in a distilled form, in this kind of learning style/representation. Go is really an awesome language and learning compilers through it while learning the language itself is a wonderful experience.

**Comment by NicoNex** [comment]
Hi, what do you mean by "standalone" binary? 
Tau is interpreted so even though you can compile it and write the bytecode to a file you still would need the Tau interpreter (or VM) to run it.

If by "standalone" you mean an ELF binary in that case Tau doesn't do that.

**Reply by geocine** [reply]
Yes like the output when you do a `go build` on golang. I understood that `tau` doesn't do that. Probably I should've asked if you have plans to do that, (for learning purposes)? 

I could definitely explore it on my own but having both your brains to pick on how you would do it, if you were to do it definitely would help someone who wants to do it. As what you have both done really helped me understand more of things.

**Reply by NicoNex** [reply]
I just used `encoding/gob` to encode a `*compiler.Bytecode` object to `[]byte` and write it to a file.
Then I still use `gob` to decode it back to `*compiler.Bytecode`.

**Reply by NicoNex** [reply]
It might not be the best solution though.

**Reply by prologic** [reply]
If you're interested in learning about Compilers that compile to "native code" as opposed to "bytecode" and interpreted by a VM, then I recommend having a look at and studying the implementations of:

- [ark-lang](https://github.com/ark-lang/ark)
- [tre](https://github.com/zegl/tre)

The later of which is a bit less code to wrap your head around. The downside however is that you need the LLVM toookkit/compiler installed whereas Go basically emits straight up machine code (it has a built-in assembler and linker)
